Dart/src/com/jetbrains/lang/dart/ide/completion/DartServerCompletionContributor.java

```diff
@@ -141,18 +139,35 @@ protected void addCompletions(@NotNull final CompletionParameters parameters,
                  }
 
                  updatedResultSet.addElement(lookupElement);
-               }, (includedSet, includedKinds, includedRelevanceTags) -> {
+               }, (includedSet, includedKinds, includedRelevanceTags, libraryFile) -> {
                  final AvailableSuggestionSet suggestionSet = das.getAvailableSuggestionSet(includedSet.getId());
                  if (suggestionSet == null) {
                    return;
                  }
 
+                 final Map<String, Map<String, Set<String>>> existingImports = das.getExistingImports(libraryFile);
                  for (AvailableSuggestion suggestion : suggestionSet.getItems()) {
                    final String kind = suggestion.getElement().getKind();
                    if (!includedKinds.contains(kind)) {
                      continue;
                    }
 
+                   Set<String> importedLibraries = new HashSet<>();
+                   for (Map.Entry<String, Map<String, Set<String>>> entry : existingImports.entrySet()) {
+                     String importedLibraryUri = entry.getKey();
+                     Map<String, Set<String>> importedLibrary = entry.getValue();
+                     // Checks whether any of the libraries exported by this import declares the same name as this suggestion.
+                     if (importedLibrary.entrySet().stream().anyMatch(
```

Comment:
I think this is very inefficient.
We don't use the fact that this is a map, and just iterate over entries, checking the key.
We should rewrite it to something like `importedElements[suggestion.uri].contains(suggestion.label)`.

This is especially scary because we do this for every suggestion.

And I need to add `uri` to `AvailableSuggestion`.

Reply:
Whoops, thanks @scheglov! Updated.